backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/scheduling/policyunits/CPUPolicyUnit.java

```diff
@@ -34,61 +40,103 @@ public class CPUPolicyUnit extends PolicyUnitImpl {
-                    messages.addMessage(vds.getId(), EngineMessage.VAR__DETAIL__NOT_ENOUGH_CORES.toString());
-                    log.debug("Host '{}' has only one core per socket. Resize and pin requires more than one core per socket", vds.getName());
-                }
+    public List<VDS> filter(SchedulingContext context, List<VDS> hosts, List<VM> vmGroup, PerHostMessages messages) {
+        List<VDS> candidates = new ArrayList<>();
+
+        boolean anyVmWithUnsetDynamicTopology = vmGroup.stream()
+                .anyMatch(vm -> VmCpuCountHelper.isResizeAndPinPolicy(vm)
+                            && !VmCpuCountHelper.isDynamicCpuTopologySet(vm));
+
+        // the number of host's shared CPUs has to >= as the number of CPUs of any VM
+        int maxVmGroupSharedCpuCount = vmGroup.stream()
+                .filter(vm -> !vm.getCpuPinningPolicy().isExclusivelyPinned())
+                .mapToInt(vm -> VmCpuCountHelper.isDynamicCpuTopologySet(vm)
+                        ? vm.getCurrentNumOfCpus(false)
+                                : vm.getNumOfCpus(false))
+                .max()
+                .orElse(0);
+
+        for (VDS host : hosts) {
+
+            Integer hostCpuCount = SlaValidator.getEffectiveCpuCores(host, context.getCluster().getCountThreadsAsCores());
+
+            if (hostCpuCount == null) {
+                candidates.add(host);
                 continue;
             }
-            Integer cores = SlaValidator.getEffectiveCpuCores(vds, context.getCluster().getCountThreadsAsCores());
-            if (cores != null) {
-                int numOfCpus = VmCpuCountHelper.isDynamicCpuTopologySet(vm) ?
-                        vm.getCurrentNumOfCpus(false) : vm.getNumOfCpus(false);
-                if (vm.getCpuPinningPolicy() == CpuPinningPolicy.DEDICATED) {
-                    int futureCpus = context.getCluster().getCountThreadsAsCores() ? cores - vds.getVmsCoresCount() :
-                            cores - (int) Math.ceil(vds.getVmsCoresCount() / (vds.getCpuThreads() / vds.getCpuCores()));
-                    if (numOfCpus > futureCpus) {
-                        messageNotEnoughCores(vds, cores, vm, messages);
-                        continue;
-                    }
-                } else {
-                    int takenCpus = 0;
-                    // takenCpus are CPUs (threads), we should consider switching it to cores when necessary.
-                    takenCpus = context.getCluster().getCountThreadsAsCores() ?
-                            vdsCpuUnitPinningHelper.getDedicatedCount(vds.getId()) :
-                            vdsCpuUnitPinningHelper.countTakenCores(vds);
-                    cores = cores - takenCpus;
-                    if (numOfCpus > cores) {
-                        messageNotEnoughCores(vds, cores, vm, messages);
-                        continue;
-                    }
+
+            // when the dynamic topology is not set, the host requires more then one core per socket
```

Comment:
and a vm is set with `resize and pin`.

Reply:
Actually, is it necessary to check if the dynamic topology is set or not? Isn't it enough to just check that if there is any VM with Resize and PIN policy in the vm group, the host has to have more than 1 core per socket?

Reply:
Yes, thinking of it. It may be redundant. The thought was first run versus migration. But in both cases the host won't fit if it's 1 core.

Reply:
I will remove the check for the dynamic topology then - that would make the code a bit more easier.